### PR TITLE
Any-to-one VC downstream

### PR DESCRIPTION
**Pull request by unilight**
This is an initial PR for merging the any-to-one VC downstream.
Please help me test if everything is working well. I documented the execution details in the README, so please simply follow the steps in it.
I did not write the instructions in `s3prl/downstream/README.md` because I think the original one is getting too long and it's really hard to read. I hope you are okay with this.

Let me know if there is anything unclear!

### CODE REVIEW

File: s3prl/downstream/a2o-vc-vcc2020/README.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+# Download the pretrained PWGs.
+./pwg_download.sh ./
+```
+
+### Training
+Here we train an A2O VC model using the `config_ar_taco2.yaml` configuration. Please replace the `<trgspk>`, `<upstream>` in the following command to your desired settings.
+```
+cd <root-to-s3prl>/downstream/a2o-vc-vcc2020
+python ../../run_downstream.py -m train --config config_ar_taco2.yaml -n a2o_vc_vcc2020_<trgspk>_<upstream> -u <upstream> -d a2o-vc-vcc2020 -o "config.trgspk='<trgspk>'"
```
Comment (leo19941227) [inline s3prl/downstream/a2o-vc-vcc2020/README.md:53]: I notice that `config.trgspk` does not appear in the config file. I am thinking will it be nice to add placeholder in the example config files? like simply `trgspk: TEF1` and a few lines describing where people can find other speaker codes.

File: s3prl/downstream/a2o-vc-vcc2020/README.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+### Preparation
+```
+# Download the VCC2020 dataset.
+cd <root-to-s3prl>/s3prl/downstream/a2o-vc-vcc2020
+cd data
+./data_download.sh vcc2020/
+cd ../
+
+# Download the pretrained PWGs.
+./pwg_download.sh ./
```
Comment (leo19941227) [inline s3prl/downstream/a2o-vc-vcc2020/README.md:46]: It seems the pwd_download.sh was not added into git tracking XD

File: s3prl/downstream/a2o-vc-vcc2020/README.md (hunk 1 above)
Comment (leo19941227) [reply s3prl/downstream/a2o-vc-vcc2020/README.md:53]: I think it will be better to put all downstream-specific configs under the `downstream_expert` field. Gathering configurations together might make `DownstramExpert` more modularized and will make future disentanglement from `run_downstream.py` script easier. Like, someday people might want all these standalone task-modules and train them jointly with another multi-task script.

File: s3prl/downstream/a2o-vc-vcc2020/expert.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+                wav_save_path = os.path.join(wav_save_dir, "_".join(wav_path.split("/")[-2:]))
+                write(
+                    wav_save_path,
+                    self.datarc["fbank_config"]["fs"],
+                    (y * np.iinfo(np.int16).max).astype(np.int16),
+                )
+
+        print(f'{split} loss: {loss:.6f}')
+
+        save_names = []
```
Comment (leo19941227) [inline s3prl/downstream/a2o-vc-vcc2020/expert.py:227]: Just confirming, so we will not save any checkpoints since you already saved all the inferenced files into expdir right?

File: s3prl/downstream/a2o-vc-vcc2020/config_ar_taco2.yaml [hunk 4]
```diff
@@ -0,0 +1,60 @@
+runner:
+  total_steps: 10000
```
Comment (leo19941227) [inline s3prl/downstream/a2o-vc-vcc2020/config_ar_taco2.yaml:2]: I have a minor question irrelevant to implementation. How much time will it take for a single run under this optimization steps?

File: s3prl/downstream/a2o-vc-vcc2020/README.md (hunk 1 above)
Comment (leo19941227) [reply s3prl/downstream/a2o-vc-vcc2020/README.md:53]: I find that actually due to the default config path, it will be more recommended to run the `run_downstream.py` script under `s3prl/s3prl` instead of `s3prl/s3prl/run_downstream.py`.

**Review by leo19941227 — CHANGES_REQUESTED** [review]
Hi @unilight!

Many thanks for the great work! I think we only need some minor changes here. I can run through the training/evaluation and it went smoothly! Also the documentation is really comprehensive and easy to follow especially the data preparation script. Thanks again!

File: s3prl/downstream/a2o-vc-vcc2020/config.yaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  name: AdamW
+  lr: 1.0e-4
+
+# comment the whole scheduler config block
+# to disable learning rate scheduling
+scheduler:
+  name: linear_schedule_with_warmup
+  num_warmup_steps: 4000
+
+downstream_expert:
```
Comment (leo19941227) [inline s3prl/downstream/a2o-vc-vcc2020/config.yaml:28]: I think it might be nice to put target speaker configuration inside this `downstream_expert` field and give an dummy example like `trgspk: TEF1  # downstream/a2o-vc-vcc2020/data/lists"
`

File: s3prl/downstream/a2o-vc-vcc2020/README.md (hunk 1 above)
Comment (unilight) [reply s3prl/downstream/a2o-vc-vcc2020/README.md:53]: >  am thinking will it be nice to add placeholder in the example config files? like simply trgspk: TEF1

Sure. I can change this.

> I think it will be better to put all downstream-specific configs under the downstream_expert field.

Sorry, I don't understand what you suggest here...

> it will be more recommended to run the run_downstream.py script under s3prl/s3prl instead of s3prl/s3prl/downstream/vc_folder.

I agree to run `run_downstream.py` in `s3prl/s3prl` if users want to train only one downstream model. 
How about `batch_vc_train.sh`? Originally I put it in `s3prl/s3prl`, but I think that kind of breaks the structure, so I moved `batch_vc_train.sh` to `s3prl/s3prl/downstream/vc_folder`. Then I thought it might be confusing if a single dry run is done in `s3prl/s3prl` but batch run is in `s3prl/s3prl/downstream/vc_folder`... Should we do it this way?

File: s3prl/downstream/a2o-vc-vcc2020/README.md (hunk 2 above)
Comment (unilight) [reply s3prl/downstream/a2o-vc-vcc2020/README.md:46]: My bad... I'll add it.

File: s3prl/downstream/a2o-vc-vcc2020/expert.py (hunk 3 above)
Comment (unilight) [reply s3prl/downstream/a2o-vc-vcc2020/expert.py:227]: We actually are saving the checkpoints, however not used in the evaluation.

File: s3prl/downstream/a2o-vc-vcc2020/config_ar_taco2.yaml (hunk 4 above)
Comment (unilight) [reply s3prl/downstream/a2o-vc-vcc2020/config_ar_taco2.yaml:2]: It depends on the upstream and GPU. Smaller upstreams like APC are fast, and large models like wav2vec2 are slow. On a RTX 3090, training one model takes 3-5 hrs.

File: s3prl/downstream/a2o-vc-vcc2020/config.yaml (hunk 5 above)
Comment (unilight) [reply s3prl/downstream/a2o-vc-vcc2020/config.yaml:28]: Sure. I'll change this.

File: s3prl/downstream/a2o-vc-vcc2020/README.md (hunk 2 above)
Comment (unilight) [reply s3prl/downstream/a2o-vc-vcc2020/README.md:46]: Ah, I found that I changed the file name. It should be `vocoder_download.sh` now. I'll change the README.

File: s3prl/downstream/a2o-vc-vcc2020/README.md (hunk 1 above)
Comment (unilight) [reply s3prl/downstream/a2o-vc-vcc2020/README.md:53]: > I think it will be better to put all downstream-specific configs under the downstream_expert field.

I think I understand now. Do you mean to put the `trgspk` field in the `downstream_expert` field, like `"config.downstream_expert.trgspk='<trgspk>'"`?

File: s3prl/downstream/a2o-vc-vcc2020/README.md (hunk 2 above)
Comment (unilight) [reply s3prl/downstream/a2o-vc-vcc2020/README.md:46]: If you did not have the PWGs, how did you test the evaluation scripts? 😮

**Comment by unilight** [comment]
Hi @leo19941227 , my main revisions are:
- change from `config.trgspk` to `config.downstream_expert.trgspk`.
- change the instruction about single dry training run in README.
- change the vocoder download part in README.

File: s3prl/downstream/a2o-vc-vcc2020/README.md (hunk 1 above)
Comment (leo19941227) [reply s3prl/downstream/a2o-vc-vcc2020/README.md]: > > I think it will be better to put all downstream-specific configs under the downstream_expert field.
> 
> I think I understand now. Do you mean to put the `trgspk` field in the `downstream_expert` field, like `"config.downstream_expert.trgspk='<trgspk>'"`?

Yes! This is what I mean :)

Comment (leo19941227) [reply s3prl/downstream/a2o-vc-vcc2020/README.md]: > I agree to run `run_downstream.py` in `s3prl/s3prl` if users want to train only one downstream model.
> How about `batch_vc_train.sh`? Originally I put it in `s3prl/s3prl`, but I think that kind of breaks the structure, so I moved `batch_vc_train.sh` to `s3prl/s3prl/downstream/vc_folder`. Then I thought it might be confusing if a single dry run is done in `s3prl/s3prl` but batch run is in `s3prl/s3prl/downstream/vc_folder`... Should we do it this way?

Oh! I can understand your concern... I think actually both ways work (people can either execute at the `s3prl/s3pr/` or inside the vc folder) and it is just about how we document it and *what is the path of the **data folder** in configs*. I think the major policy is that I would like to keep the *benchmarking* usage consistent. That is, the specific setting for benchmarking new representations is able to run from `s3prl/s3prl` as default so that people can run through all the tasks at `s3prl/s3prl`. In your case I think it is fine to put more functions inside the vc folder which is great! I will suggest that we can divide the README into these parts:

1. How to benchmark new upstream with VC task with a simple command. (execute at `s3prl/s3prl`)
2. How to do VC-related experiments provided by this VC folder. (execute at VC folder)

We can even provide different config files with different default path of the **data folder**.
Do you think we can go this way?

File: s3prl/downstream/a2o-vc-vcc2020/README.md (hunk 2 above)
Comment (leo19941227) [reply s3prl/downstream/a2o-vc-vcc2020/README.md]: No! I not yet test the evaluate or decode scripts. What I meant by running through the train/evaluation is actually just the training/evaluation loop defined in the expert.py :) In your case the evaluation loop simply dump the files for future evaluation so I not yet touch the vocoder and metric computation parts

File: s3prl/downstream/a2o-vc-vcc2020/expert.py (hunk 3 above)
Comment (leo19941227) [reply s3prl/downstream/a2o-vc-vcc2020/expert.py:227]: OK I find it. You save the latest 10 checkpoints.

File: s3prl/downstream/a2o-vc-vcc2020/config_ar_taco2.yaml (hunk 4 above)
Comment (leo19941227) [reply s3prl/downstream/a2o-vc-vcc2020/config_ar_taco2.yaml:2]: Cool this is fast enough!

**Review by leo19941227 — COMMENTED** [review]
Hi @unilight, I also ran through the decoding part and it works!
So I think after we resolve the remaining comments we are ready to merge this :)

File: s3prl/downstream/a2o-vc-vcc2020/README.md (hunk 1 above)
Comment (unilight) [reply s3prl/downstream/a2o-vc-vcc2020/README.md]: > I think it is great to put more functions inside the vc folder and I will suggest that we can divide the README into two parts:

> We can even provide different config files with different default path of the data folder. Do you think we can go this way?

Sure! That sounds perfect to me! I think I will create 4 config files:

- `config.yaml`: the default config to benchmark a new upstream, which contains **relative** paths, so needs to be executed in `s3prl/s3prl`.
- `config_simple.yaml`, `config_simple_ar.yaml` and `config_taco2_ar.yaml`: the configs for reproducing experiments in my own VC paper, which contains **absolute** paths, so can be executed in the VC folder.

File: s3prl/downstream/a2o-vc-vcc2020/decode.sh [hunk 6]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
+# *********************************************************************************************
+#   FileName     [ decode.sh ]
+#   Synopsis     [ neural vocoder decoding & objective evaluation script for voice conversion ]
+#   Author       [ Wen-Chin Huang (https://github.com/unilight) ]
+#   Copyright    [ Copyright(c), Toda Lab, Nagoya University, Japan ]
+# *********************************************************************************************
+
+voc_expdir=$1
```
[the remark this answers is not in the document]
Comment (unilight) [reply s3prl/downstream/a2o-vc-vcc2020/decode.sh:9]: Sure! I'll change this.

File: s3prl/downstream/a2o-vc-vcc2020/README.md (hunk 1 above)
Comment (leo19941227) [reply s3prl/downstream/a2o-vc-vcc2020/README.md]: Cool! This also sounds perfect to me.

**Review by leo19941227 — APPROVED** [review]

**Comment by leo19941227** [comment]
Hi @unilight,

The changes look good to me. I am merging this PR.
Thanks again for the great work!

**Comment by unilight** [comment]
@leo19941227  I am still writing the README! Can you wait?

**Comment by leo19941227** [comment]
Sure! I was waiting for you to confirm before the merging.
Let me know if you are ready!

**Comment by unilight** [comment]
@leo19941227 I finished modifying the README in the VC directory. BTW, do I also need to modify `s3prl/downstream/README.md`?

**Comment by leo19941227** [comment]
I will prefer to add the minimal information about benchmarking new representations with VC task on the downstream/README.md. Like one line for data preparation, one line for training, one line for deciding etc (just for example, not restricted to one line.)

**Comment by unilight** [comment]
Cool! I'll work on that now.

**Comment by unilight** [comment]
@leo19941227 Done!
